File: packages/camera/camera_android_camerax/test/camera_info_test.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:camera_android_camerax/src/camera_info.dart';
import 'package:camera_android_camerax/src/camera_state.dart';
import 'package:camera_android_camerax/src/camerax_library.g.dart';
import 'package:camera_android_camerax/src/instance_manager.dart';
import 'package:camera_android_camerax/src/live_data.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/annotations.dart';
import 'package:mockito/mockito.dart';

import 'camera_info_test.mocks.dart';
import 'test_camerax_library.g.dart';

@GenerateMocks(<Type>[
  LiveData<CameraInfo>,
  TestCameraInfoHostApi,
  TestInstanceManagerHostApi
])
void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  // Mocks the call to clear the native InstanceManager.
  TestInstanceManagerHostApi.setup(MockTestInstanceManagerHostApi());

  group('CameraInfo', () {
    tearDown(() => TestCameraInfoHostApi.setup(null));

    test('getSensorRotation makes call to retrieve expected sensor degrees',
        () async {
      final MockTestCameraInfoHostApi mockApi = MockTestCameraInfoHostApi();
      TestCameraInfoHostApi.setup(mockApi);

      final InstanceManager instanceManager = InstanceManager(
        onWeakReferenceRemoved: (_) {},
      );
      final CameraInfo cameraInfo = CameraInfo.detached(
        instanceManager: instanceManager,
      );
      instanceManager.addHostCreatedInstance(
        cameraInfo,
        0,
        onCopy: (_) => CameraInfo.detached(),
      );

      when(mockApi.getSensorRotationDegrees(
              instanceManager.getIdentifier(cameraInfo)))
          .thenReturn(90);
      expect(await cameraInfo.getSensorRotationDegrees(), equals(90));

      verify(mockApi.getSensorRotationDegrees(0));
    });

    test('getLiveCameraState makes call to retrieve live camera state',
        () async {
      final MockTestCameraInfoHostApi mockApi = MockTestCameraInfoHostApi();
      TestCameraInfoHostApi.setup(mockApi);

      final InstanceManager instanceManager = InstanceManager(
        onWeakReferenceRemoved: (_) {},
      );
      final CameraInfo cameraInfo = CameraInfo.detached(
        instanceManager: instanceManager,
      );
      const int cameraIdentifier = 55;
      final MockLiveData<CameraState> mockLiveCameraState =
          MockLiveData<CameraState>();
      final MockLiveData<CameraState> castedMockLiveCameraState =
          MockLiveData<CameraState>();
      const int liveCameraStateIdentifier = 73;
      instanceManager.addHostCreatedInstance(
        cameraInfo,
        cameraIdentifier,
        onCopy: (_) => CameraInfo.detached(),
      );
      instanceManager.addHostCreatedInstance(
        mockLiveCameraState,
        liveCameraStateIdentifier,
        onCopy: (_) => MockLiveData<CameraState>(),
      );

      when(mockApi.getLiveCameraState(cameraIdentifier))
          .thenReturn(liveCameraStateIdentifier);
      when(mockLiveCameraState.cast<CameraState>())
          .thenReturn(castedMockLiveCameraState);

      expect(await cameraInfo.getLiveCameraState(),
          equals(castedMockLiveCameraState));
      verify(mockApi.getLiveCameraState(cameraIdentifier));
    });

    test('flutterApi creates CameraInfo instance with expected identifier', () {
      final InstanceManager instanceManager = InstanceManager(
        onWeakReferenceRemoved: (_) {},
      );
      final CameraInfoFlutterApi flutterApi = CameraInfoFlutterApiImpl(
        instanceManager: instanceManager,
      );

      flutterApi.create(0);

      expect(
          instanceManager.getInstanceWithWeakReference(0), isA<CameraInfo>());
    });
  });
}

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/CameraInfoTest.java
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.camerax;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import androidx.camera.core.CameraInfo;
import androidx.camera.core.CameraState;
import androidx.lifecycle.LiveData;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugins.camerax.GeneratedCameraXLibrary.LiveDataSupportedType;
import java.util.Objects;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

public class CameraInfoTest {
  @Rule public MockitoRule mockitoRule = MockitoJUnit.rule();

  @Mock public CameraInfo mockCameraInfo;
  @Mock public BinaryMessenger mockBinaryMessenger;

  InstanceManager testInstanceManager;

  @Before
  public void setUp() {
    testInstanceManager = InstanceManager.create(identifier -> {});
  }

  @After
  public void tearDown() {
    testInstanceManager.stopFinalizationListener();
  }

  @Test
  public void getSensorRotationDegrees_makesCallToRetrieveSensorRotationDegrees() {
    final CameraInfoHostApiImpl cameraInfoHostApi =
        new CameraInfoHostApiImpl(mockBinaryMessenger, testInstanceManager);

    testInstanceManager.addDartCreatedInstance(mockCameraInfo, 1);

    when(mockCameraInfo.getSensorRotationDegrees()).thenReturn(90);

    assertEquals((long) cameraInfoHostApi.getSensorRotationDegrees(1L), 90L);
    verify(mockCameraInfo).getSensorRotationDegrees();
  }

  @Test
  public void getLiveCameraState_makesCallToRetrieveLiveCameraState() {
    final CameraInfoHostApiImpl cameraInfoHostApiImpl =
        new CameraInfoHostApiImpl(mockBinaryMessenger, testInstanceManager);
    final LiveDataFlutterApiWrapper mockLiveDataFlutterApiWrapper =
        mock(LiveDataFlutterApiWrapper.class);
    final Long mockCameraInfoIdentifier = 27L;
    @SuppressWarnings("unchecked")
    final LiveData<CameraState> mockLiveCameraState = (LiveData<CameraState>) mock(LiveData.class);

    testInstanceManager.addDartCreatedInstance(mockCameraInfo, mockCameraInfoIdentifier);
    cameraInfoHostApiImpl.liveDataFlutterApiWrapper = mockLiveDataFlutterApiWrapper;
    when(mockCameraInfo.getCameraState()).thenReturn(mockLiveCameraState);

    final Long liveCameraStateIdentifier =
        cameraInfoHostApiImpl.getLiveCameraState(mockCameraInfoIdentifier);

    verify(mockLiveDataFlutterApiWrapper)
        .create(eq(mockLiveCameraState), eq(LiveDataSupportedType.CAMERA_STATE), any());
    assertEquals(
        liveCameraStateIdentifier,
        testInstanceManager.getIdentifierForStrongReference(mockLiveCameraState));
  }

  @Test
  public void flutterApi_makesCallToDartToCreateInstance() {
    final CameraInfoFlutterApiImpl spyFlutterApi =
        spy(new CameraInfoFlutterApiImpl(mockBinaryMessenger, testInstanceManager));

    spyFlutterApi.create(mockCameraInfo, reply -> {});

    final long identifier =
        Objects.requireNonNull(testInstanceManager.getIdentifierForStrongReference(mockCameraInfo));
    verify(spyFlutterApi).create(eq(identifier), any());
  }
}
